Update Project Manager Protocol

radeusgd opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

Closes #1149 

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

radeusgd commented:
Some notes on the current state of protocol update:

I tried to make the update backwards compatible:
- There are added message types, but they do not have to be used yet.
- The `project/open` and `create` were modified, but by adding an optional field, so it is compatible.
- New error types were added.

So I will be able to create infrastructure for the new protocol version that will just ignore the unimplemented features and the IDE does not have to change the protocol yet, being able to wait until the features are actually usable.

## Installing Components

I propose the following way of handling missing components:
- The IDE can by default send an open/create request with `installMissingComponents = false`.
- If the engine is installed, it will just do the thing.
- If the engine is not installed, it will result in a specific error: `MissingComponentError`.
- The IDE can check for this error, and for example ask the user '$errorMessage \n Do you want to install the missing engine?'.
- If the user answers affirmative, the request is sent again, this time with `installMissingComponents = true`, triggering installation of the missing component.
  - (Or if by some coincidence it has been installed in the meantime, it will just proceed with opening).

## Progress Bars 

The installation may take some time, so the CLI displays progress bars, at least for some of these longer steps.

Each progress bar has a lifecycle of first being initialized with `task/started` (which contains a UUID for the task), being updated with `task/progress-update` and finalized with `task/finished`.

`task/finished` may include an error (but please note that regardless of the task-related error, the error will also be reported for the original request associated with the task, for example as `ComponentInstallationError` returned for the `project/open` request that triggered the installation).

`task/progress-update` includes a `done` value, which depending on context may mean for example processed bytes and an optional `total` value which indicates for example the total expected bytes.
If `total` is present, it can be used to compute a percentage to show. `total` may be missing if the total cannot be estimated (like when downloading a file without size specified in headers or just waiting for a lock to be released) - in this case some indeterminate progress bar can be displayed.

### Multiple Sub-Tasks

Currently the installation consists of two to four parts that may display a progress bar:
- downloading the engine package
- extracting the engine package
- downloading and extracting the runtime if it was not installed yet

The concept of progress bars is taken directly from the CLI and so these progress bars are treated independently, i.e. with the current architecture the communication will be following:

- -> `project/open` request
- <- `task/started` (download)
- many <- `task/progress-update` (download)
- <- `task/finished` (download)
- <- `task/started` (extracting)
- many <- `task/progress-update` (extracting)
- <- `task/finished` (extracting)
- <- `project/open` response

So while all progress updates happen within the `project/open` response-request (up to possible reordering of messages), the two sub-tasks of download and extract are completely independent.

Is that okay or do we want to somehow connect them by creating a notion of sub-tasks?
I would suggest staying with that is there for now, as the progress bars can simply show up and disappear, a bit like progress bars associated to background tasks in IntelliJ and if needed the grouping may be added later to keep things simple for now, but I would like to ask for feedback on that.

### Progress Unit

Most of the time `done` means processed bytes, but that is not always the case. If the client knows the unit, it may use that to compute things like throughput in bytes per second. Do you think that the progress updates should include the unit (for now a union of `bytes` or `unknown`) making it more meaningful to the client or should that be left out for a later extension?

iamrecursion commented:
I don't have any opinions on the protocol itself, but the documentation is very high quality!

farmaazon commented on docs/language-server/protocol-project-manager.md:
````diff
@@ -185,10 +230,34 @@ This message requests the creation of a new project.
 }
 ```
 
+> TODO [RW] discuss: Do we also want a 'latest-available' that would check the
+> repository for newer versions? While we are at it, as the version needs to be
+> specified somehow, we may want to be able to query all available version
+> strings from the repository? To allow users to choose a version for their new
+> project, besides just `engine/list` which only lists currently installed
+> versions.
+
````
My proposition is to make engine/list return also possible-to-install versions with appropriate mark.

radeusgd replied:
I'd suggest having a separate endpoint, `engine/list-available` (and possibly renaming this one to `engine/list-installed`), as for example if the user is offline they may still want to list the installed engines to uninstall one of them, but listing available versions will in such case fail with a network error. So I think keeping them separate will be simpler.

radeusgd replied:
Also I understand that in one way or another the IDE can get a list of all available versions, the `latest-available` option for `version` in `project/create` is not needed because it can be inferred from this list.

In the same way, maybe the `latest-instaled` should also be removed since we can list the installed version, leaving just a specific semver string or missing the field to indicate the default version?

farmaazon replied:
> Also I understand that in one way or another the IDE can get a list of all available versions, the latest-available option for version in project/create is not needed because it can be inferred from this list.
> 
> In the same way, maybe the latest-instaled should also be removed since we can list the installed version, leaving just a specific semver string or missing the field to indicate the default version?

Correct

> I'd suggest having a separate endpoint, engine/list-available (and possibly renaming this one to engine/list-installed), as for example if the user is offline they may still want to list the installed engines to uninstall one of them, but listing available versions will in such case fail with a network error. So I think keeping them separate will be simpler.

I'm ok with that.

farmaazon commented on docs/language-server/protocol-project-manager.md:
```diff
@@ -304,22 +381,377 @@ interface ProjectListSampleResponse {
+  /** Indicates amount of progress, for example: count of processed bytes. */
+  done: Int;
+
+  /** Indicates total expected progress.
+   *
+   * May be missing, as it is not always known, for example when downloading a
+   * file of unknown size or waiting on a lock.
+   */
+  total?: Int;
```
I think this should go to `task/started`. it should never change during progress, and we will immediately know if we have to display normal, or indeterminate progress bar.

radeusgd replied:
It shouldn't change indeed, but the issue is that the total progress may not be available immediately when the task is started - for example when downloading a file we may not know its size before we actually start the request and receive the headers.

But as that is an implementation detail, so I see no harm in simply postponing the `task/started` message until the total is known (i.e. in the current implementation, until the first progress update).

farmaazon replied:
> But as that is an implementation detail, so I see no harm in simply postponing the task/started message until the total is known (i.e. in the current implementation, until the first progress update).

Yes, actually we can't do much with single "task/start" message without any hint what progress bar we ought to display.

farmaazon reviewed (commented):
> Is that okay or do we want to somehow connect them by creating a notion of sub-tasks?
I would suggest staying with that is there for now, as the progress bars can simply show up and disappear, a bit like progress bars associated to background tasks in IntelliJ and if needed the grouping may be added later to keep things simple for now, but I would like to ask for feedback on that.

I agree with your suggestion.

> Do you think that the progress updates should include the unit (for now a union of bytes or unknown) making it more meaningful to the client or should that be left out for a later extension?

Yes, especially when the task don't know the total value. However I prefer "none" and/or "other" instead of "unknown" (what does it mean we don't know unit)?

> I have proposed to treat this as a normal error but it has one major downside: If I understand correctly, errors cannot contain additional fields besides code and message (or am I wrong and I can add new fields to a specific error type?) and so there is no place to include which version the project manager should be upgraded to.

Well, on our side we are prepared for additional fields in Errors, so adding them should be fine

> if a newer version is required to install a specific engine, the launcher upgrades not to the minimum required version but just to the latest one.

However IDE need to have ability to quick fallback to older versions, and thus knowing the minimal required version for this project may be very useful.

radeusgd commented on docs/language-server/protocol-project-manager.md:
```diff
@@ -185,10 +230,34 @@ This message requests the creation of a new project.
 interface ProjectCreateRequest {
+  /** Name of the project to create. */
   name: String;
+
+  /** Enso Engine version to use for the project.
+   *
+   * Possible values are:
+   * - a semver version string identifying an Enso engine version,
+   * - `default` to use the current default
+   * - `latest-installed` to use the most recent version that is installed on
+   *   the system.
+   */
+  version: String;
+
```
Ok, it will just be `default` if left out.

radeusgd commented on docs/language-server/protocol-project-manager.md:
````diff
@@ -304,22 +381,377 @@ interface ProjectListSampleResponse {
+```typescript
+interface TaskStartNotification {
+  taskId: UUID;
+  taskName: String;
````
Hmm, I actually haven't thought through what exactly the name should represent.
My initial idea was that it would be some label related to the specific task (i.e. `downloading`, `extracting`), but if the method name is more helpful for you, I think it can be implemented without issue.

radeusgd replied:
In this case I think I will rename this field into `relatedOperation` maybe?

radeusgd commented:
I have updated the documentation according to the discussion, highlights:

- `version` is optional for compatibility.
- Added better explanation for the 'task' notification workflow.
- Added the unit field, currently allowing `bytes` or `other`.
- Moved total to `task/started` and changed `taskName` to `relatedOperation`.
- The errors may include an additional `data` field, the upgrade error contains `data.minimumRequiredVersion`.

I think that the only thing left to discuss is how we want to handle `engine/list`.

radeusgd commented on docs/language-server/protocol-project-manager.md:
````diff
@@ -82,6 +109,12 @@ specified project.
 ```typescript
 interface ProjectOpenRequest {
   projectId: UUID;
+
+  /** Specifies whether to install missing components.
+   *
+   * If not provided, defaults to false.
+   */
+  installMissingComponents?: bool;
````
One more thing that we may need to discuss.

Apart from a component just being missing, the version that is required may be missing *and* marked as broken. The launcher does not allow installing `broken` versions by default, unless forced and I guess the IDE should also display some kind of warning about the version being broken and suggesting to upgrading to a more recent version.

My suggestion is to add an analogous `forceInstallBrokenComponents?: bool` that will default to `false` and will normally be set as such. Then if the missing engine is marked as broken the operation could fail with `MissingComponentIsBroken` error and that can be recovered by the IDE in a similar way missing components are.

Analogous parameter may or may not be needed for `project/create`. Alternatively, create could proceed with broken versions, as before selecting the version to create, you need to find that this version even exists and `engine/list-available` can include that information. Or maybe the broken releases should be completely ignored for purposes of creating new projects?

farmaazon replied:
> My suggestion is to add an analogous forceInstallBrokenComponents?: bool that will default to false and will normally be set as such. Then if the missing engine is marked as broken the operation could fail with MissingComponentIsBroken error and that can be recovered by the IDE in a similar way missing components are.

Ok, however one enum with three variants is better than two bools. Or with two variants, as this is optional field, as you please

Concerning `project/create`: I would keep the same option as in `project/open`, for simplicity. IDE may just don't use all variants, or e.g. guard creating project with broken versions with three irritating warning dialogs.

farmaazon reviewed (approved):
LGTM, but I didn't check scala

4e6 commented on docs/language-server/protocol-project-manager.md:
```diff
@@ -304,22 +424,492 @@ interface ProjectListSampleResponse {
+  relatedOperation: String;
+
+  /** Unit in which progress of this task is measured. */
+  unit: "bytes" | "other";
```
Should be a separate ADT type, like `MissingComponentAction`

4e6 commented on lib/scala/json-rpc-server/src/main/scala/org/enso/jsonrpc/Protocol.scala:
```diff
@@ -103,8 +107,13 @@ object Errors {
   case object ServiceError   extends Error(1, "Service error")
-  case class UnknownError(override val code: Int, override val message: String)
-      extends Error(code, message)
+  case object NotImplementedError
```
Also, add to the error descriptions in docs
